library/core/tests/array.rs

```diff
@@ -28,11 +28,22 @@ fn array_try_from() {
         ($($N:expr)+) => {
             $({
                 type Array = [u8; $N];
-                let array: Array = [0; $N];
+                let mut array: Array = [0; $N];
                 let slice: &[u8] = &array[..];
 
                 let result = <&Array>::try_from(slice);
                 assert_eq!(&array, result.unwrap());
+
+                let result = <Array>::try_from(slice);
+                assert_eq!(&array, &result.unwrap());
+
+                let mut_slice: &mut [u8] = &mut array[..];
+                let result = <&mut Array>::try_from(mut_slice);
+                assert_eq!(&[0; $N], result.unwrap());
+
+                let mut_slice: &mut [u8] = &mut array[..];
+                let result = <Array>::try_from(mut_slice);
+                assert_eq!(&array, &result.unwrap());
```

Comment:
nit: Why is this assert different from the one above? They're expecting the exact same `result` AFAICT.

Reply:
Do you mean the left hand side of `assert_eq` at L42 (`&[0; $N]`) is different from one at this line (`&array`)?

I used `&[0; $N]` at L42 because `&array` is not available there. `array` is exclusively borrowed at L40. So it is not accessible while `result` is alive.

Reply:
Hmm, the borrow on L40 should no longer be live after L41. Can you show me the error you get with using `array` on L42? Otherwise, perhaps this is related to #89966.

Reply:
> the borrow on L40 should no longer be live after L41.

Since `mut_slice` is converted into `&mut [u8; N]` and the mutable reference is stored at `result` at L41, it's still alive at L42. (`array` → `mut_slice` → `result`)

Here is an error message I got when I replaced `&[0; $N]` with `&array`:

```
error[E0502]: cannot borrow `array` as immutable because it is also borrowed as mutable
  --> library/core/tests/array.rs:42:28
   |
27 | /     macro_rules! test {
28 | |         ($($N:expr)+) => {
29 | |             $({
30 | |                 type Array = [u8; $N];
...  |
40 | |                 let mut_slice: &mut [u8] = &mut array[..];
   | |                                                 ----- mutable borrow occurs here
41 | |                 let result = <&mut Array>::try_from(mut_slice);
42 | |                 assert_eq!(&array, result.unwrap());
   | |                            ^^^^^^  ------ mutable borrow later used here
   | |                            |
   | |                            immutable borrow occurs here
...  |
48 | |         }
49 | |     }
   | |_____- in this expansion of `test!`
50 | /     test! {
51 | |          0  1  2  3  4  5  6  7  8  9
52 | |         10 11 12 13 14 15 16 17 18 19
53 | |         20 21 22 23 24 25 26 27 28 29
54 | |         30 31 32
55 | |     }
   | |_____- in this macro invocation

For more information about this error, try `rustc --explain E0502`.
error: could not compile `core` due to previous error
```

Reply:
Ah, I see 👍 

I misread the `<&mut Array>::try_from` as going *from* a `&mut Array`, rather than into.